### ISSUE TITLE
Make tensorflow and tensorflow-devel deb packages

### ISSUE THREAD

**Issue by ghost** [tensorflow/tensorflow#720]
Will you make it possible to install and update it by distributions package systems (deb & rpm)? You can make an official repository.
Will you make it possible to install a Tensorflow-devel package, which gives acces to C++ headers to make programs, which link to Tensorflow?

**Comment by Sherif89** [comment]
You'd be able to dynamically link Tensorflow, while building it,  to libraries available by the distribution.

**Comment by martinwicke** [comment]
We want to provide at least apt and homebrew packages. If someone would like to help out, we'd be thrilled.

**Comment by Sherif89** [comment]
Also, C++ applications'd be able to use the latest version without having to recompile.

**Comment by martinwicke** [comment]
Isn't that only true if we update the binary packages? Which realistically,
we'd only do for releases.

**Comment by Sherif89** [comment]
Yes, correct me, if wrong. If you dynamically link to a C++ library and the library is updated, you'll use the latest version of this library.

**Comment by christopherowen** [comment]
+1 for a yum repo containing this.

**Comment by bhack** [comment]
Debian [ITP #804612](https://bugs.debian.org/804612) blocked by Bazel [ITP #782654](https://bugs.debian.org/782654)

**Comment by CDLuminate** [comment]
@bhack  You're right.

Things like splitting `libtensorflow` and `libtensorflow-devel` are done by the distributions instead of the upstream.

In order to get package `Tensorflow` into Debian's official archive, currently we need first to get `bazel` into the archive. However, if another build system is provided (Tensorflow #380 ), tensorflow gets rid of the blocker `bazel`.

As I'm holding Debian's ITP for tensorflow, I'm waiting for either an alternative build system, or the bazel being packaged and uploaded.

**Comment by Sherif89** [comment]
Would it be necessary to add another package for Python support? You can also add a doc package.

**Comment by bhack** [comment]
In debian Bazel it is blocked by protobuf >= 3.0.0 [ITP #795841](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=795841).

**Comment by Sherif89** [comment]
What about Arch, OpenSUSE or Fedora?

**Comment by bhack** [comment]
See [Arch](https://aur.archlinux.org/packages/?O=0&SeB=nd&K=tensorflow&outdated=&SB=n&SO=a&PP=50&do_Search=Go)

**Comment by Sherif89** [comment]
> See Arch

Can it be added to a repository to be installable by pacman? Why does it use Python 2?

**Comment by CDLuminate** [comment]
You can install these arch AUR packages with `yaourt`, see arch wiki for detail.

Arch packages (`pkgbuild`) always require less working hour on packaging, so you may find it very fast that arch community adopts new packages.

So please just be patient about the distro's progress, many contributors are working on it under  a non-profit circumstance.

**Comment by Sherif89** [comment]
I'm very patient. I just hopehope it's part of its roadmap to make it:
1. Easy to install & update like any package.
2. Easy to link with C++ programs and the latest Python version.

**Comment by CDLuminate** [comment]
I'm sure that Debian will have at least `CUDA7.5+` and `Caffe` officially in this year (before or within Q2) but not sure for `Tensorflow`. And then Ubuntu devs will pull these Debian packages into Ubuntu vivid. As for other distros I'm afraid I can provide no info.

`1` must be in the roadmap. 
`2` is a bit complicated ....

**Comment by CDLuminate** [comment]
Dirty hacked packages come out very fast.

**Comment by Sherif89** [comment]
What's necessary to package Tensorflow?
Would #1309 make it closer?

**Comment by martinwicke** [comment]
#1309 would solve the bazel issue for Debian.

**Comment by CDLuminate** [comment]
OK, it would be great. When #1309 is merged, I can unblock tensorflow and then continue.

**Comment by ricochet2200** [comment]
When #1309 is fixed, I'll try out the cmake build on CentOS 7.  CentOS 7 doesn't have JDK 1.8 or bazel in its yum repo so using bazel to roll the rpm would be cumbersome.  Supposedly bazel works with 1.7 even though it is deprecated, but it wouldn't work for me.  

Rolling rpms should be much more straightforward after we can build with cmake.

**Comment by iahmad-khan** [comment]
@ricochet2200 any update? i just need a spec file that can build tensor-flow. 
centos6 , amd-64

**Comment by ricochet2200** [comment]
@iahmad-khan I haven't been following this bug closely.  Can you build with cmake now?

**Comment by martinwicke** [comment]
Just the core for now -- but it may still be interesting to make a tensorflow-core package from that.

**Comment by Sherif89** [comment]
You can put it in the open build service. It supports many distributions (Ubuntu, Fedora, OpenSUSE,...).

**Comment by gunan** [comment]
@martinwicke could you comment?

**Comment by martinwicke** [comment]
It will become more interesting as we fix the problems with our linkage / symbol visibility. But for now, I'm reluctant to maintain more packages.

I would still welcome contributions here.

**Comment by lukeyeager** [comment]
Getting Tensorflow uploaded to the official debian repository is currently blocked because Bazel doesn't follow the Debian policy manual (https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=782654#45). But you could still provide deb packages even if they aren't accepted to upstream Debian.

If you are able to provide source packages for Tensorflow and all of its dependencies, then you could upload the packages to a PPA. Despite the bazel dependency, it sounds like this might be possible according to https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=782654#65.

* DIGITS ([scripts](https://github.com/NVIDIA/DIGITS/tree/83254ed7a0/packaging/deb/templates), [build](https://github.com/NVIDIA/DIGITS/blob/83254ed7a09/packaging/deb/Dockerfile#L30-L32), [upload](https://github.com/NVIDIA/DIGITS/blob/83254ed7a09/scripts/travis/ppa-upload.sh#L25))
* Ethereum ([scripts](https://github.com/ethereum/go-ethereum/tree/c008176f9f/build), [build+upload](https://github.com/ethereum/go-ethereum/blob/c008176f9f/build/ci.go#L456))

Or, if you can't provide source packages for tensorflow or one of it's dependencies (e.g. if you're working on a CUDA build), then you could still build your own and host it on your own servers (like you currently do for Python wheels).

* nvidia-docker ([scripts](https://github.com/NVIDIA/nvidia-docker/tree/v1.0.1/build/deb), [build](https://github.com/NVIDIA/nvidia-docker/blob/v1.0.1/Dockerfile.deb#L32-L37))

**Comment by CDLuminate** [comment]
@lukeyeager The most straightforward way to make a tensorflow deb package is to repack the official whl package as a deb package, in which the Bazel dependency is skipped.

Tensorflow packaging is in my todolist but with a low priority, since pip / conda already provided a quite convenient way to install. Currently some Debian developer is also interested in packaging tensorflow, and he is currently trying to solve the Bazel ITP.

**Comment by lukeyeager** [comment]
@CDLuminate: the TF package would be simple, but I realized today I would have to re-package a bunch of its dependencies, too. It looked like I was going to have to make new packages for `protobuf`, `weakref`, `bleach`, `html5lib`, `markdown`, and `werkzeug` (and possibly some of the sub-dependencies of those packages) in order to satisfy all the dependencies on 16.04. Yuck.

**Comment by martinwicke** [comment]
Most of the js deps will go away (tensorboard is moving to a separate pip
package) so it could be somewhat simpler.

**Comment by CDLuminate** [comment]
@lukeyeager New Debian stable had just been released. Developers will upload new versions of their packages soon. AFAIK, for instance, the protobuf 3.3.1 packaging is already available but still not uploaded. Some new packages still stay in experimental.

**Comment by dmacvicar** [comment]
https://github.com/tensorflow/tensorflow/issues/13061

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by mehditlili** [comment]
@cdluminate  I see you already started doing [some work](https://salsa.debian.org/science-team/tensorflow) on the packaging, is it already packageable? I tried running  dpkg-deb --build tensorflow but still get errors. An updated Readme would help.

**Comment by cdluminate** [comment]
@mehditlili Well, without error log I cannot tell you anything useful.

salsa/science-team/tensorflow
---
This repo holds the CPU version. For detail about progress please lookup README.Debian .

I can build the package from the git repo you referred on at least two architectures (amd64,s390x). The latest result is that the package was successfully compiled on amd64 and x390x. Some other architectures failed to build due to OOM killer during parallel `ld` so I don't know if the package compiles on them.

https://launchpad.net/~lumin0/+archive/ubuntu/ppa/+build/15309497
https://launchpad.net/~lumin0/+archive/ubuntu/ppa/+build/15309502

Besides, my targeting distribution is Debian experimental. However, it's impossible to write any documentation for user at this stage because this package is a really experimental WIP and there are still many pending choices.

**Comment by sylvestre** [comment]
@cdluminate I would be happy to sponsor it in Debian if you need!

**Comment by cdluminate** [comment]
@sylvestre I reuploaded the source to PPA with the `-gsplit-dwarf` flag and let's see what will be the difference. Thanks for the hint!

**Comment by mehditlili** [comment]
@cdluminate I am trying to convert it into a  Ubuntu 16.04 package. Did you encounter any dependency problem when building this package, was that a reason why it was only done for Ubuntu 18?

**Comment by cdluminate** [comment]
@mehditlili I'm moving to a new build system written from scratch with python plus ninja. And currently the debian directory is a bit messy. Please don't try to debug this messy WIP until I updated debian/rules for the new python+ninja build system.

As for dependency, yes indeed many system-shipped packages in Ubuntu 16.04 is too old to build tensorflow. If you encounter a build failure, try to embed a copy of newer source code to the package and use it. From my PPA you see only package for Ubuntu 18 is built because I don't care about old releases but only **Debian unstable/experimental**.

**Comment by cdluminate** [comment]
@mehditlili My build system is nearly in shape. Is it working for you?

https://lists.debian.org/debian-mentors/2018/09/msg00044.html

**Comment by cdluminate** [comment]
@sylvestre I forgot to CC you in the latest mail.

Preliminary binary packages available here:

http://debomatic-amd64.debian.net/debomatic/experimental/pool/tensorflow_1.10.1+dfsg-A1u27/tensorflow_1.10.1+dfsg-A1u27_amd64.changes

readme: https://lists.debian.org/debian-mentors/2018/09/msg00052.html

**Comment by harshini-gadige** [comment]
Closing the issue as per the solution provided above. Can reopen the issue if it still exists.

**Comment by cdluminate** [comment]
Yes this issue has been resolved. The first upload of TF is currently waiting in Debian's upload queue.

**Comment by harshini-gadige** [comment]
> Yes this issue has been resolved. The first upload of TF is currently waiting in Debian's upload queue.

Great !!

**Comment by ytrezq** [comment]
> Closing the issue as per the solution provided above. Can reopen the issue if it still exists.

@Harshini-Gadige looks like you can reopen.
Especially since I am completely failing to get it running on aarch64 which means only a binary package would solve the problem.
As a result, I can t run spleeter from Deezer.

**Comment by advancedwebdeveloper** [comment]
Any progress for Ubuntu 20.04?

**Comment by Dilshat** [comment]
Is there any light in the tunnel?

**Comment by victorbocharov** [comment]
@Dilshat  I'm also waiting for. Here is the latest information about packages in Debian (Removed from experimental in 2019-08-27):
https://tracker.debian.org/pkg/tensorflow
https://salsa.debian.org/deeplearning-team/tensorflow/

**Comment by FrancisMurtagh-arm** [comment]
https://tracker.debian.org/news/1301497/accepted-tensorflow-231-1-source-amd64-into-experimental-experimental/

2.3.1 re-entered experimental thanks to Wookey@debian.org
